src/main/java/com/google/devtools/build/lib/packages/SelectorList.java

```diff
@@ -132,15 +133,25 @@ static SelectorList of(Iterable<?> values) throws EvalException {
    * Creates a list that concatenates two values, where each value may be a native type, a select
    * over that type, or a selector list over that type.
    *
-   * @throws EvalException if the values don't have the same underlying type
+   * @throws IllegalStateException if the values don't have the same underlying type
    */
-  static SelectorList concat(Object x, Object y) throws EvalException {
+  static SelectorList concat(Object x, Object y) {
     return of(Arrays.asList(x, y));
   }
 
   @Override
   public SelectorList binaryOp(TokenKind op, Object that, boolean thisLeft) throws EvalException {
-    if (op == TokenKind.PLUS) {
+    if (!canConcatenate(getNativeType(this), getNativeType(that))) {
+      throw Starlark.errorf(
+          "'%s' operator applied to incompatible types (%s, %s)",
+          op.toString(), getTypeName(thisLeft ? this : that), getTypeName(thisLeft ? that : this));
+    }
+    if (getNativeType(that).equals(Dict.class)) {
+      if (op == TokenKind.PIPE) {
+        return thisLeft ? concat(this, that) : concat(that, this);
+      }
+    }
```

Comment:
This means we are pointlessly checking canConcatenate twice: in `binaryOp` and in `of()`.

A solution, IMHO, would be to make `concat` a private method which simply does `return new SelectorList(getNativeType(x), ImmutableList.of(x, y))` without performing any `canConcatenate` check. It should be easy to do since we're no longer calling `concat` from SelectorValue.

Reply:
Tried making `concat` private and return a new SelectorList directly but
  - A few tests use `concat`, so the visibility cannot be private, it would require using `@VisibileForTesting`
  - `of` flattens the SelectorLists elements, which we would not do here instantiating the SelectorList directly, causing errors like `expected type <type> for attribute <attr>, but got select`

Instead I removed the check entirely so we rely on `of` to fail if the types cannot be combined.